middleware/request/retry.go

```diff
@@ -8,21 +8,33 @@ import (
 )
 
 // NewRetryMiddleware returns a new retry middleware that wraps the provided `next` request handler.
-// The middleware retries the request a maximum of `maxRetries` times with a delay of `interval` between each retry.
+// The middleware accepts 2 params consisting of:
+//   - `getRetryInterval`: A higher order function that returns a function to calculate next retry interval
+//   - `shouldRetry`: A higher order function that returns a function to calculate if we should retry again or not
+//
 // If the request succeeds at any retry, the middleware returns `nil`. If all retries fail, it returns the last error encountered.
-func NewRetryMiddleware(maxRetries int, interval time.Duration) func(next wasabi.RequestHandler) wasabi.RequestHandler {
+func NewRetryMiddleware(getRetryInterval GetRetryInterval, shouldRetry ShouldRetry) func(next wasabi.RequestHandler) wasabi.RequestHandler {
 	return func(next wasabi.RequestHandler) wasabi.RequestHandler {
 		return dispatch.RequestHandlerFunc(func(conn wasabi.Connection, req wasabi.Request) error {
 			var err error
-			ticker := time.NewTicker(interval)
+
+			ticker := time.NewTicker(getRetryInterval(0))
+
 			defer ticker.Stop()
-			for i := 0; i < maxRetries; i++ {
+
+			for {
+				condition, i := shouldRetry()
```

Comment:
This looks like unnecessary complication in current implementation, it would make sense if would give additional inputs to the function as for example last error message, and we can apply some custom logic to it.

Reply:
@ksysoev then lets keep maxRetries as a param. Because it feels counter intuitive to me that `getRetryInterval` will return `duration = 0`. The client will have to read through Retry Middleware code to understand that it expects the get interval function to return 0 for break condition. Should we keep max retries as a param for now? #94 is likely to change this

Reply:
ready for review again